src/lib.rs

```diff
@@ -116,40 +117,60 @@ pub struct Rules {
 }
 
 impl Rules {
-    /// Split a text into segments.
-    pub fn split<'a>(&self, text: &'a str) -> Vec<&'a str> {
+    /// Obtain the ranges for text segments. Guaranteed to be at character bounds.
+    pub fn split_ranges(&self, text: &str) -> Vec<Range<usize>> {
         let mut segments = Vec::new();
 
-        let mut mask: Vec<Option<bool>> = vec![None; text.len()];
+        // Mask defines the segments of text which are
+        // TODO use a proper tri-state enum here
```

Comment:
"which are..." what? :slightly_smiling_face: 

I agree regarding the tri-state enum. But it's not important; we can leave this as a `TODO`.